File: WarmUp/searchclient/searchclient/Node.java
package searchclient;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.LinkedList;
import java.util.Random;

import searchclient.Command.Type;

public class Node {
	private static final Random RND = new Random(1);

	public static int MAX_ROW = 70;
	public static int MAX_COL = 70;

	// Arrays are indexed from the top-left of the level, with first index being row and second being column.
	// Row 0: (0,0) (0,1) (0,2) (0,3) ...
	// Row 1: (1,0) (1,1) (1,2) (1,3) ...
	// Row 2: (2,0) (2,1) (2,2) (2,3) ...
	// ...
	// (Start in the top left corner, first go down, then go right)
	// E.g. this.walls[2] is an array of booleans having size MAX_COL.
	// this.walls[row][col] is true if there's a wall at (row, col)
	//

	public static boolean[][] walls = new boolean[MAX_ROW][MAX_COL];
	public static ArrayList<Goal> goals = new ArrayList<>();
	public ArrayList<Agent> agents = new ArrayList<>();
	// public Box[][] boxes = new Box[MAX_ROW][MAX_COL];
	public ArrayList<Box> boxes = new ArrayList<>();

	public Node parent;
	public Command action;

	private int g;
	
	private int _hash = 0;

	public Node(Node parent) {
		this.parent = parent;
		if (parent == null) {
			this.g = 0;
		} else {
			this.g = parent.g() + 1;
		}
	}

	public int g() {
		return this.g;
	}

	public boolean isInitialState() {
		return this.parent == null;
	}

	public boolean isGoalState()
	{
		for(Goal goal : goals)
		{
			if(!goal.goalState)
				return false;
		}
		return true;
	}

	public ArrayList<Node> getExpandedNodes()
	{
		ArrayList<Node> expandedNodes = new ArrayList<Node>(Command.EVERY.length);
		for(Agent a : agents)
		{
			int agentRow = a.location.getLeft();
			int agentCol = a.location.getRight();
            System.err.println("AgentRow: " + String.valueOf(agentRow));
			System.err.println("AgentCol: " + String.valueOf(agentCol));
			for (Command c : Command.EVERY)
			{
			    System.err.println("Action: " + c.toString());
				// Determine applicability of action
				int newAgentRow = agentRow + Command.dirToRowChange(c.dir1);
				int newAgentCol = agentCol + Command.dirToColChange(c.dir1);

				if (c.actionType == Type.Move)
				{
				    System.err.println("newAgentRow: " + String.valueOf(newAgentRow));
				    System.err.println("newAgentCol: " + String.valueOf(newAgentCol));
					// Check if there's a wall or box on the cell to which the agent is moving
					if(this.cellIsFree(newAgentRow, newAgentCol))
					{
						Node n = this.ChildNode(this.boxes, this.agents);
						n.action = c;
						n.getAgent(agentRow, agentCol).setLocation(new Pair<>(newAgentRow, newAgentCol));
						expandedNodes.add(n);
					}
				}
				else if (c.actionType == Type.Push)
				{
					// Make sure that there's actually a box to move
					if (this.boxAt(newAgentRow, newAgentCol))
					{
						int newBoxRow = newAgentRow + Command.dirToRowChange(c.dir2);
						int newBoxCol = newAgentCol + Command.dirToColChange(c.dir2);
						// .. and that new cell of box is free
						if (this.cellIsFree(newBoxRow, newBoxCol))
						{
                            Node n = this.ChildNode(this.boxes, this.agents);
                            n.action = c;
							n.getAgent(agentRow, agentCol).setLocation(new Pair<>(newAgentRow, newAgentCol));
							n.getBox(newAgentRow, newAgentCol).setLocation(new Pair<>(newBoxRow, newBoxCol));
							// n.boxes[newBoxRow][newBoxCol] = this.boxes[newAgentRow][newAgentCol];
							// n.boxes[newAgentRow][newAgentCol] = 0;
							expandedNodes.add(n);
						}
					}
				}
				else if (c.actionType == Type.Pull)
				{
					// Cell is free where agent is going
					if (this.cellIsFree(newAgentRow, newAgentCol))
					{
						int boxRow = agentRow + Command.dirToRowChange(c.dir2);
						int boxCol = agentCol + Command.dirToColChange(c.dir2);
						// .. and there's a box in "dir2" of the agent
						if (this.boxAt(boxRow, boxCol))
						{
                            Node n = this.ChildNode(this.boxes, this.agents);
                            n.action = c;
							n.getAgent(agentRow, agentCol).setLocation(new Pair<>(newAgentRow, newAgentCol));
							n.getBox(boxRow, boxCol).setLocation(new Pair<>(agentRow, agentCol));
							// n.boxes[this.agentRow][this.agentCol] = this.boxes[boxRow][boxCol];
							// n.boxes[boxRow][boxCol] = 0;
							expandedNodes.add(n);
						}
					}
				}
			}
		}
		Collections.shuffle(expandedNodes, RND);
		return expandedNodes;
	}

	private Box getBox(int row, int col)
	{
		for(Box box : this.boxes)
		{
			if(box.location.getLeft() == row && box.location.getRight() == col)
				return box;
		}
		throw new NullPointerException("No box at row: " + String.valueOf(row) + " and col: " + String.valueOf(col));
	}

	private Goal getGoal(int row, int col)
	{
		for(Goal goal : goals)
		{
			if(goal.location.getLeft() == row && goal.location.getRight() == col)
				return goal;
		}
		throw new NullPointerException("No goal at row: " + String.valueOf(row) + " and col: " + String.valueOf(col));
	}

	private Agent getAgent(int row, int col)
	{
		for(Agent a : this.agents)
		{
			if(a.location.getLeft() == row && a.location.getRight() == col)
				return a;
			else
            {
                System.err.println("Row: " + a.location.getLeft());
                System.err.println("Col: " + a.location.getRight());
            }
		}
		throw new NullPointerException("No agent at row: " + String.valueOf(row) + " and col: " + String.valueOf(col));
	}

	private boolean cellIsFree(int row, int col)
	{
		return !walls[row][col] && !boxAt(row, col) && !agentAt(row, col);
	}

	private boolean boxAt(int row, int col)
	{
		for(Box box : this.boxes)
		{
			if(box.location.getLeft() == row && box.location.getRight() == col)
				return true;
		}
		return false;
	}

	private boolean goalAt(int row, int col)
	{
		for(Goal goal : goals)
		{
			if(goal.location.getLeft() == row && goal.location.getRight() == col)
				return true;
		}
		return false;
	}

	private boolean agentAt(int row, int col)
	{
		for(Agent a : this.agents)
		{
			if(a.location.getLeft() == row && a.location.getRight() == col)
				return true;
		}
		return false;
	}

	private Node ChildNode(ArrayList<Box> boxes, ArrayList<Agent> agents) {
		Node copy = new Node(this);
		copy.boxes = boxes;
		copy.agents = agents;
		return copy;
	}

	public LinkedList<Node> extractPlan() {
		LinkedList<Node> plan = new LinkedList<Node>();
		Node n = this;
		while (!n.isInitialState()) {
			plan.addFirst(n);
			n = n.parent;
		}
		return plan;
	}

	@Override
	public int hashCode() {
		if (this._hash == 0) {
			final int prime = 31;
			int result = 1;
			result = prime * result + this.agents.hashCode();
			result = prime * result + this.boxes.hashCode();
			result = prime * result + goals.hashCode();
			result = prime * result + Arrays.deepHashCode(this.walls);
			this._hash = result;
		}
		return this._hash;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (this.getClass() != obj.getClass())
			return false;
		Node other = (Node) obj;
		if(this.agents.equals(other.agents))
			return false;
		//if (!Arrays.deepEquals(this.boxes, other.boxes))
		if(!this.boxes.equals(other.boxes))
			return false;
		// if (!Arrays.deepEquals(this.goals, other.goals))
			// return false;
		if (!Arrays.deepEquals(this.walls, other.walls))
			return false;
		return true;
	}

	@Override
	public String toString() {
		StringBuilder s = new StringBuilder();
		for (int row = 0; row < MAX_ROW; row++) {
			if (!this.walls[row][0]) {
				break;
			}
			for (int col = 0; col < MAX_COL; col++)
			{
				if (boxAt(row, col)) {
					s.append(getBox(row, col).chr);
				} else if (goalAt(row, col)) {
					s.append(getGoal(row, col).chr);
				} else if (this.walls[row][col]) {
					s.append("+");
				} else if (agentAt(row, col)) {
					s.append(getAgent(row, col).num);
				} else {
					s.append(" ");
				}
			}
			s.append("\n");
		}
		return s.toString();
	}

}

File: ProgrammingProject/searcher/src/Node.cpp
#include "Node.h"
#include <string>
#include <vector>
#include <list>
#include <ctype.h>
#include <algorithm>
#include <random>
#include <functional>
#include <iostream>
#include <iomanip>
#define RANDOM_SEED 1
#include "Goal.h"
#include "Agent.h"

bool Node::operator==(const Node * obj) const{
	return (equals(obj));
}

	//Initialize static variables:
	int Node::maxX;
	int Node::maxY;
	std::vector<bool> Node::walls;
	//Maybe make it true goals instead of pointers, as this one never needs editing.
	std::vector<Goal *> Node::goals;

	//This should only be used for the first node. Inits walls & goals.

	Node::Node() {
		this->parent = NULL;
		this->gval = 0;
		this->boxes = std::vector<Box *>();
	}



	Node::Node(Node * parent) {
		this->parent = parent;
		this->gval = parent->g() + 1;
		//Inits boxes
		this->agents = DeepCloneAgents(parent->agents);
		this->boxes = DeepCloneBoxes(parent->boxes);
	}

	int Node::g() {
		return this->gval;
	}

	bool Node::isInitialState() {
		return this->parent == NULL;
	}

	std::vector<Node *> Node::getExpandedNodes(){
		std::vector<Node *> expandedNodes = std::vector<Node *>();
		std::cerr << "3.1\n";
		for (Agent * a : agents){
			std::cerr << "3.2\n";
			for (Command * c : Command::EVERY) {
				// Determine applicability of action
				int newAgentX = a->getX() + Command::dirToXChange(c->dir1);
				int newAgentY = a->getY() + Command::dirToYChange(c->dir1);

				if (c->actionType == Command::Move) {

					// Check if there's a wall or box on the cell to which the agent is moving
					if (this->cellIsFree(newAgentX, newAgentY)) {
						std::cerr << "3.3\n";

						Node * n = this->ChildNode();
						n->action = c;
						n->getAgent(a->getX(), a->getY())->setLocation(newAgentX, newAgentY);
						expandedNodes.push_back(n);
					}
				} else if (c->actionType == Command::Push) {
					std::cerr << "3.4\n";
					// Make sure that there's actually a box to move
					if (this->boxAt(newAgentX, newAgentY)) {
						int newBoxX = newAgentX + Command::dirToXChange(c->dir2);
						int newBoxY = newAgentY + Command::dirToYChange(c->dir2);
						// .. and that new cell of box is free
						if (this->cellIsFree(newBoxX, newBoxY)) {
							std::cerr << "3.5\n";

							Node * n = this->ChildNode();
							std::cerr << "3.5\n";

							n->action = c;
							n->getAgent(a->getX(), a->getY())->setLocation(newAgentX, newAgentY);
							std::cerr << "3.6\n";

							n->getBox(newAgentX, newAgentY)->setLocation(newBoxX, newBoxY);
							std::cerr << "3.7\n";

							expandedNodes.push_back(n);
							std::cerr << "3.8\n";
						}
					}
				} else if (c->actionType == Command::Pull) {
					// Cell is free where agent is going
					if (this->cellIsFree(newAgentX, newAgentY)) {
						int boxX = a->getX() + Command::dirToXChange(c->dir2);
						int boxY = a->getY() + Command::dirToYChange(c->dir2);
						// .. and there's a box in "dir2" of the agent
						if (this->boxAt(boxX, boxY)) {

							Node * n = this->ChildNode();
							n->action = c;
							n->getAgent(a->getX(), a->getY())->setLocation(newAgentX, newAgentY);
							n->getBox(newAgentX, newAgentY)->setLocation(a->getX(), a->getY());
							expandedNodes.push_back(n);
						}
					}
				}
			}
		}


		return expandedNodes;
	}


	Node * Node::ChildNode() {
		std::cerr << "3.10\n";

		Node * copy = new Node(this);
		//This works because std::vector. Copies full 1D-array. Thank god for 1D :)
		copy->boxes = this->boxes;
		return copy;
	}

	std::list<Node *> Node::extractPlan() {
		std::list<Node*> plan = std::list<Node*>();
		Node * n = this;

		while (!(n->isInitialState())) {
			plan.push_front(n);
			n = n->parent;
		}
		return plan;
	}

	int Node::hashCode() const{

			int prime = 31;
			int result = 1;
			for (Agent * a : agents){
				result = prime * result + a->getX();
				result = prime * result + a->getY();
			}
			//Constructs a string and hashes it. Easier than hashing random vectors.
			for (Box * b : boxes){
				result = prime * result + b->getX();
				result = prime * result + b->getY();
			}

			return result;
	}

	bool Node::equals(const Node * obj) const {
		if (obj == NULL)
			return false;
			//Using == works with std::vector
			//Assumes the size of agents is the same. Maybe this should be rectified later.
			//This might be an issue. We must ensure that the order of boxes and agents are always identical.
			for (int i = 0; i < agents.size(); i++){
				if (!agents[i]->equals(obj->agents[i])){
					return false;
				}
			}

			//Assumes the size of boxes is the same. Maybe this should be rectified later.
			//This might be an issue. We must ensure that the order of boxes and agents are always identical.
			for (int i = 0; i < boxes.size(); i++){
				if (!boxes[i]->equals(obj->boxes[i])){
					return false;
				}
			}

		return true;
	}

	std::string Node::toString() {
		std::string s("");

			for (int y = 0; y < maxY; y++) {
				for (int x = 0; x < maxX; x++) {
					if (Node::walls[x+y*maxX]) {
						s += ("+");
					} else {
						s += (" ");
					}
				}
			s.append("\n");
			}
			//Write a goal at location for each goal.
			for (Goal * g : goals){
				int x = g->getX();
				int y = g->getY();
				s[x+y*(maxX+1)] = g->chr;
			}
			//Write a box at location for each goal.
			for (Box * b : boxes){
				int x = b->getX();
				int y = b->getY();
				s[x+y*(maxX+1)] = b->chr;
			}

			for (Agent * a : agents){
				int x = a->getX();
				int y = a->getY();
				s[x+y*(maxX+1)] = (char) (a->num + (int) '0');
			}


		return s;
	}

	bool Node::isGoalState()
	{
		for(Goal * goal : Node::goals)
		{
			bool goalState = false;
			for(Box * box : this->boxes){
				if(goal->location == box->location){
					goalState = true;
					break;
				}
			}
		if(!goalState)
			return false;
		}
		return true;
	}

	Box * Node::getBox(int x, int y)
	{
		std::pair<int, int> loc(x,y);
		for(Box * box : this->boxes)
		{
			if(box->getLocation() == loc)
				return box;
		}
		return NULL;
	}

	Goal * Node::getGoal(int x, int y)
	{
		std::pair<int, int> loc(x,y);
		for(Goal * goal : goals)
		{
			if(goal->getLocation() == loc)
				return goal;
		}
		//throw new NullPointerException("No goal at row: " + String.valueOf(row) + " and col: " + String.valueOf(col));
		return NULL;
	}


	Agent * Node::getAgent(int x, int y)
	{
		std::pair<int, int> loc(x,y);
		for(Agent * a : this->agents)
		{
			if(a->getLocation() == loc)
				return a;
		}
		return NULL;
		//throw new NullPointerException("No agent at row: " + String.valueOf(row) + " and col: " + String.valueOf(col));
	}

	bool Node::cellIsFree(int x, int y)
	{
		return !walls[x + y*maxX] && !boxAt(x, y) && !agentAt(x, y);
	}

	bool Node::boxAt(int x, int y)
	{
		std::pair<int, int> loc(x,y);
		for(Box * box : this->boxes)
		{
			if(box->getLocation() == loc)
				return true;
		}
		return false;
	}

	bool Node::goalAt(int x, int y)
	{
		std::pair<int, int> loc(x,y);
		for(Goal * goal : goals)
		{
			if(goal->getLocation() == loc)
				return true;
		}
		return false;
	}

	bool Node::agentAt(int x, int y)
	{
		std::pair<int, int> loc(x,y);
		for(Agent * a : this->agents)
		{
			if(a->getLocation() == loc)
				return true;
		}
		return false;
	}

	std::vector<Agent*> Node::DeepCloneAgents(std::vector<Agent*> agents)
	{
		std::vector<Agent *> clone = std::vector<Agent *>(agents.size());
		for(int i = 0; i < agents.size(); i++){
			(clone[i]) = new Agent(agents[i]);
		}
		return clone;
	}

	std::vector<Box*> Node::DeepCloneBoxes(std::vector<Box *> boxes)
	{
		std::vector<Box *> clone = std::vector<Box *>(boxes.size());
		for(int i = 0; i < boxes.size(); i++){
			(clone[i]) = new Box(boxes[i]);
		}
		return clone;
	}
